File: src/main/java/dev/alexengrig/leetcode/_347_top_k_frequent_elements/HeapSolution.java
package dev.alexengrig.leetcode._347_top_k_frequent_elements;

import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Queue;

class HeapSolution implements Task {
    @Override
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> frequencyMap = new HashMap<>();
        for (int num : nums) {
            frequencyMap.compute(num, (key, count) -> count == null ? 1 : count + 1);
        }
        Queue<Map.Entry<Integer, Integer>> heap = new PriorityQueue<>(Comparator.comparingInt(Map.Entry::getValue));
        for (Map.Entry<Integer, Integer> entry : frequencyMap.entrySet()) {
            heap.offer(entry);
            if (heap.size() > k) {
                heap.poll();
            }
        }
        int[] results = new int[k];
        for (int i = 0; i < k && !heap.isEmpty(); i++) {
            results[i] = heap.poll().getKey();
        }
        return results;
    }
}

File: src/main/python/347_top_k_frequent_elements/heap_solution.py
import heapq
from collections import Counter
from typing import List

from task import Task


class HeapSolution(Task):
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counts = Counter(nums)
        heap = []
        for num, count in counts.items():
            heapq.heappush(heap, (count, num))
            if len(heap) > k:
                heapq.heappop(heap)
        return [num for count, num in heap]
